### PR TITLE
Bos xdoc attn

### PR DESCRIPTION
**Pull request by Alvorecer721**
1. Remove all EOD-related code from the `_get_ltor_masks_and_position_ids` function. 
2. Add BOS-related code and slightly improve code quality.
3. Add a `--no-cross-document-attention` option to the data configuration to disable cross-document attention.
4. BOS token hiding regardless of whether cross-document attention is active.

### CODE REVIEW

**Review by mpagli — APPROVED** [review]
I did not try to run the code, but the logic in `_get_ltor_masks_and_position_ids` makes sense to me. Thanks for all the nice comments. 

LGTM.

File: megatron/core/datasets/gpt_dataset.py [hunk 1]
```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
...
+        bos_indices = (data == bos_token).nonzero(as_tuple=True)[0]
+
+        # Loop through BOS indices:
+        if bos_indices.numel() > 0:  # Only process if BOS tokens exist (codewise not strictly required but improves clarity.)
+            prev_index = 0
+            for i in bos_indices: 
+                if attention_mask is not None:
+                    # BOS token masking
+                    attention_mask[0, :, i] = 0  # No tokens can attend to BOS
+                    attention_mask[0, i, :] = 0  # BOS can't attend to any tokens
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:696]: I am confused by these two lines. Why are you masking BOS tokens? And if you are, why are you even keeping them? These tokens will add unnecessary compute (negligible, but still) but will have no effect on the model's processing. 

Also, even if you are, shouldn't this be done only if "reset_attention_mask" is true?

File: megatron/core/datasets/gpt_dataset.py [hunk 2]
```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
...
+                        # Block cross document attention
+                        attention_mask[0, i:, :i+1] = 0
+
+                        # If cross document attention is blocked, mask loss for last token of previous document
+                        if i > 0:  # Skip first BOS as there's no previous document
+                            loss_mask[i-1] = 0.0  # Mask loss for token right before BOS
+                    
+                # Reset positions.
+                if reset_position_ids:
+                    if i > 0:  # Skip first BOS as it's already at position 0
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:708]: 1. if the formula is correct the additional if makes it more complicated but is not needed. you'll be basically doing - 0 (and it also works as a sanity check to make sure everything is working as expected). 
2. I am not fully familiar with the codebase but usually position ids start from 0. In that case, I think the formula should be (i - prev_index - 1) and prev_index should be set to -1 at the beginning. This doesn't matter much for RoPE though since it's relative but would be good to fix.

File: megatron/core/datasets/gpt_dataset.py [hunk 3]
```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
...
+            prev_index = 0
+            for i in bos_indices: 
+                if attention_mask is not None:
+                    # BOS token masking
+                    attention_mask[0, :, i] = 0  # No tokens can attend to BOS
+                    attention_mask[0, i, :] = 0  # BOS can't attend to any tokens
+                    
+                    if reset_attention_mask:
+                        # Block cross document attention
+                        attention_mask[0, i:, :i+1] = 0
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:700]: the 3rd index should be :i. It's relevant to my above comment:  to me it seems like you shouldn't be masking BOS, but even if you are, the place for doing it is not here and per the comment this should only block cross-doc attention.

File: megatron/core/datasets/gpt_dataset.py [hunk 4]
```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
         position_ids = position_ids.clone()
 
     if reset_position_ids or reset_attention_mask:
-        # Find indices where EOD token is.
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:672]: Are we dropping support for EODs?

File: megatron/core/datasets/gpt_dataset.py [hunk 5]
```diff
@@ -92,6 +95,10 @@ def __init__(
         super().__init__(
             indexed_dataset, dataset_path, indexed_indices, num_samples, index_split, config
         )
+        if not self.config.cross_document_attention:
+            self.config.reset_position_ids = True
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:99]: I am not fully familiar with the code base to know whether this is something widely done or not. but since this is an argument in the config, I'd suggest to add asserts here instead of directly changing it. Changing it like this can lead to a lot of confusion later on, where people see they have set this to False in the config and yet, it somehow ends up being true.

File: megatron/core/datasets/gpt_dataset.py (hunk 3 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:700]: Yes I realized this, I had a discussion with Imanol and we will make change accordingly. Thanks for pointing out.

File: megatron/core/datasets/gpt_dataset.py (hunk 4 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:672]: Yes because we don't have EOD in our data

File: megatron/core/datasets/gpt_dataset.py (hunk 1 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:696]: We are studying how BOS tokens affect model memorisation. We have already trained and evaluated models where BOS tokens can actively participate in attention. Now, we need to compare this with models where BOS tokens serve purely as structural markers for document boundaries, without contributing to the semantic understanding. 

This is achieved by:
- Keeping BOS tokens in the input (to mark document starts)
- Preventing them from influencing the model's processing through attention masking
- Ensuring they're never predicted as outputs

This allows us to isolate and study the impact of BOS tokens on memorization by comparing models with 'active' vs 'structural-only' BOS tokens.

Comment (mkrima) [reply megatron/core/datasets/gpt_dataset.py:696]: ok then should this be a config flag to set whether we are using active vs structural BOS? also since you are merging this upstream already, is this a test or the decision has been made that all future runs (except experimental ones with "active" BOS tokens) should use structural-only BOS tokens?

File: megatron/core/datasets/gpt_dataset.py [hunk 6]
```diff
@@ -619,6 +632,7 @@ def _build_shuffle_index(
 
 def _get_ltor_masks_and_position_ids(
     data: torch.Tensor,
+    bos_token: int,
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:635]: also, maybe add this new argument to the docstring of the function.

File: megatron/core/datasets/gpt_dataset.py [hunk 7]
```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
...
+                        attention_mask[0, i:, :i+1] = 0
+
+                        # If cross document attention is blocked, mask loss for last token of previous document
+                        if i > 0:  # Skip first BOS as there's no previous document
+                            loss_mask[i-1] = 0.0  # Mask loss for token right before BOS
+                    
+                # Reset positions.
+                if reset_position_ids:
+                    if i > 0:  # Skip first BOS as it's already at position 0
+                        position_ids[i:] = position_ids[i:] - (i - prev_index)
```
Comment (mkrima) [inline megatron/core/datasets/gpt_dataset.py:709]: I removed this comment by mistake sorry for sending it again.

Basically the correct formula here I think it i - prev_index - 1 since usually position_ids start from 0 but i'm not 100% sure about this for this codebase.

File: megatron/core/datasets/gpt_dataset.py (hunk 3 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:700]: I wrote a simple function to illustrate the logic, I will make change accordingly to the codebase.
```
def show_attention_masks(data, bos_mask=False):
    seq_length = len(data)
    bos_token = 1 
    eod_token = 2

    # Convert data to readable labels
    token_labels = ['BOS' if t == bos_token else 'EOD' if t == eod_token else 'tok' for t in data]

    def print_matrix(matrix, title):
        print(f"\n{title}")
        # Header rows
        print("         ", end="")  # Extra space for alignment
        print(" ".join(f"{d:^4}" for d in data))
        print("         ", end="")   # Extra space for alignment
        print(" ".join(f"{l:^4}" for l in token_labels))
        print("         " + "-" * (len(token_labels) * 5))  # Increased width of separator
        
        # Matrix rows
        for i in range(len(token_labels)):
            # Left side labels with fixed width
            print(f"{data[i]:2} {token_labels[i]:^3} |", end=" ")
            # Matrix values with fixed width
            for j in range(seq_length):
                print(f"{matrix[0,i,j]:^4.1f}", end=" ")
            print()

    # Initial attention mask (lower triangular)
    attention_mask = torch.tril(torch.ones((seq_length, seq_length))).unsqueeze(0)
    print_matrix(attention_mask, "Initial attention mask:")

    position_ids = torch.arange(seq_length)
    final_mask = attention_mask.clone()

    # Our BoS masking
    if bos_mask:
        bos_index = position_ids[data == bos_token]
        for j in range(len(bos_index)):
            i = bos_index[j]
            final_mask[0, i:, :i+1] = 0
    else:
        bos_index = position_ids[data == bos_token]
        for j in range(len(bos_index)):
            i = bos_index[j]
            final_mask[0, i:, :i] = 0

    print_matrix(final_mask, "Our Cross Doc BOS masking:")

data = torch.Tensor([1, 3, 3, 3, 1, 3, 3])
mask = show_attention_masks(data, bos_mask=True)
```

File: megatron/core/datasets/gpt_dataset.py (hunk 7 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:709]: Let's analyse step by step with this example where BOS=1:
```
data =            [1, 3, 3, 3, 1, 3, 3]   # BOS at positions 0 and 4
initial_pos_ids = [0, 1, 2, 3, 4, 5, 6]

# First BOS (i=0, prev_index=0):
# Skips because i == 0

# Second BOS (i=4, prev_index=0):
Using i - prev_index = 4 - 0 = 4:
pos_ids[4:] -= 4
[0, 1, 2, 3, 0, 1, 2]   # Positions reset to 0 at second BOS ✅

Using i - prev_index - 1 = 4 - 0 - 1 = 3:
pos_ids[4:] -= 3
[0, 1, 2, 3, 1, 2, 3]   # Positions reset to 1 at second BOS ❌
```
Therefore I still I think `position_ids[i:] = position_ids[i:] - (i - prev_index)` should be correct

Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:709]: The suggested `-1` would make positions start at 1 from BOS

Comment (mkrima) [reply megatron/core/datasets/gpt_dataset.py:709]: very sorry. you are right. my bad ! all good 👍

**Comment by ischlag** [comment]
Thanks for reviewing @mkrima 

just a bit of context:

Yixuen is adding features regarding memorisation. In short, bos tokens can increase memorisation. So we want to implement a flag that allows us to control if the BoS token is visible or not to the model. For simplicity, we do this here in the attention and don't remove it. 

In general, all features we add should be disabled by default. 

And since this is a general purpose codebase I think we need to be a bit careful about dropping EoD support. We do not have EoD in our own tokenisation. Only BoS. 

You can find other ablations for our model here: https://github.com/swiss-ai/pretrain/tree/main/megatron

File: megatron/core/datasets/gpt_dataset.py (hunk 3 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:700]: What `attention_mask[0, i:, :i+1] = 0` does:

1. mask out bos tokens in attention equivalent to 
   ```attention_mask[0, :, i] = 0  # No tokens can attend to BOS
        attention_mask[0, i, :] = 0  # BOS can't attend to any tokens```

File: megatron/core/datasets/gpt_dataset.py (hunk 5 above)
Comment (martinjaggi) [reply megatron/core/datasets/gpt_dataset.py]: agree!

File: megatron/core/datasets/gpt_dataset.py [hunk 8]
```diff
@@ -35,6 +35,9 @@ class GPTDatasetConfig(BlendedMegatronDatasetConfig):
     eod_mask_loss: bool = None
     """Option to enable the EOD mask loss"""
 
+    bod_hiding: bool = None
+    """Option to enbale the BOD mask attention"""
```
Comment (martinjaggi) [inline megatron/core/datasets/gpt_dataset.py:39]: -> """to enable hiding (masking) the BoD token from all attention. default = false"""

File: megatron/core/datasets/gpt_dataset.py [hunk 9]
```diff
@@ -632,18 +653,22 @@ def _build_shuffle_index(
 
 def _get_ltor_masks_and_position_ids(
     data: torch.Tensor,
-    bos_token: int,
+    bod_token: int,
```
Comment (martinjaggi) [inline megatron/core/datasets/gpt_dataset.py:656]: there is probably no reason to rename it? we'd prefer to stay as compatible as ever possible with the official megatron upstream code

File: megatron/core/datasets/gpt_dataset.py (hunk 4 above)
Comment (martinjaggi) [reply megatron/core/datasets/gpt_dataset.py:672]: could still keep EOD support as much untouched as possible (even if not used with our data for now)

File: megatron/core/datasets/gpt_dataset.py (hunk 9 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:656]: To align with Megatron’s upstream code—which uses `eod` to represent what is essentially an `eos` token ([reference](https://github.com/swiss-ai/Megatron-LM/blob/0bd247ece4ccac2a47b4b31eac75a9a0aeefd11e/megatron/training/tokenizer/tokenizer.py#L175))—I propose renaming `bos` to `bod`. Since our approach only adds the `bos` token at the start of each document rather than at every sequence boundary, using `bod` (“beginning of document”) would better capture its actual role.

Comment (martinjaggi) [reply megatron/core/datasets/gpt_dataset.py:656]: all good! sorry i thought the renaming here was compared to upstream megatron, but actually [upstream](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py) has nothing, so it's indeed best if you call it `BoD`. well done!

File: megatron/core/datasets/gpt_dataset.py (hunk 6 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py]: Resolved in [latest](https://github.com/swiss-ai/Megatron-LM/pull/40/commits/3685a9e1785daf1444eae120878c078663a1093c) commit.

File: megatron/core/datasets/gpt_dataset.py (hunk 5 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py]: Resolved in [latest](https://github.com/swiss-ai/Megatron-LM/pull/40/commits/3685a9e1785daf1444eae120878c078663a1093c) commit.

File: megatron/core/datasets/gpt_dataset.py (hunk 4 above)
Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:672]: In the upstream code, eod tokens are used to mark document boundaries, while our tokenisation relies solely on bod. This creates an awkward situation where we need to handle how the upstream code expects eod to be handled. There are several unresolved questions about handling different scenarios: 
- [x] data with only bod tokens. 
- [ ] data with only eod tokens.
- [ ] the complex case of data containing both token types.
- [ ] understand how eod relates to memorisation.
- [ ] whether we should apply the same attention masking to eod as we do with bod.

To keep things manageable, we're trying to minimise code changes - when neither `--bod-hiding` nor `--no-cross-document-attention` is used, we want to preserve the original upstream behavior so ongoing ablations aren't affected. That said, there is one change we really need to make unconditionally: the model should never learn to predict bod tokens. Right now we're controlling this with -`-bod-hiding`, but this should really be a universal change in the future.

Due to time constraints, I've implemented the simplest workable solution: keeping `eod_mask_loss` but using bod as the primary document boundary. While I agree we should maintain eod support, we need to work through all these open questions first.

File: megatron/core/datasets/gpt_dataset.py (hunk 8 above)
Comment (ischlag) [reply megatron/core/datasets/gpt_dataset.py:39]: this is the way megatron implements args. The default behaviour is relegated to the arguments.py file where this flag by default is false.

**Review by ischlag — APPROVED** [review]
lgtm
